File: src/main/java/net/atomique/ksar/XML/OSConfig.java
package net.atomique.ksar.XML;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Iterator;

public class OSConfig {

    public OSConfig(String s) {
        OsName = s;
    }

    public void addStat(StatConfig s) {
        StatHash.put(s.getStatName(), s);
    }

    public void addGraph(GraphConfig s) {
        GraphHash.put(s.getName(), s);
    }

    public String getOsName() {
        return OsName;
    }

    public String getStat(String[] columns, int firstdatacolumn) {

        final String[] result = {null};

        String[] s1 = Arrays.copyOfRange(columns, firstdatacolumn, columns.length);
        String header = String.join(" ", s1);

        StatHash.keySet().forEach((String item) -> {
            StatConfig tmp = StatHash.get(item);
            if (tmp.check_Header(header, s1.length)) result[0] = tmp.getGraphName();
        });
        return result[0];

    }

    public StatConfig getStat(String statName) {
        if (StatHash.isEmpty()) {
            return null;
        }
        Iterator<String> ite = StatHash.keySet().iterator();
        while (ite.hasNext()) {
            String tmptitle = ite.next();
            StatConfig tmp = (StatConfig) StatHash.get(tmptitle);
            if ( tmp.getGraphName().equals(statName)) {
                return tmp;
            }
        }
        return null;
    }

    public GraphConfig getGraphConfig(String s) {
        if (GraphHash.isEmpty()) {
            return null;
        }
        return GraphHash.get(s);
    }

    public HashMap<String, StatConfig> getStatHash() {
        return StatHash;
    }

    public HashMap<String, GraphConfig> getGraphHash() {
        return GraphHash;
    }
    private String OsName = null;
    private HashMap<String, StatConfig> StatHash = new HashMap<>();
    private HashMap<String, GraphConfig> GraphHash = new HashMap<>();
}

File: src/main/kotlin/net/atomique/ksar/xml/OSConfig.kt
/*
 * Copyright 2018 The kSAR Project. All rights reserved.
 * See the LICENSE file in the project root for more information.
 */
package net.atomique.ksar.xml

import java.util.Arrays
import java.util.HashMap
import java.util.function.Consumer

class OSConfig(var osName: String) {
    val statHash = HashMap<String, StatConfig>()
    val graphHash = HashMap<String, GraphConfig>()
    private var mappingStatHeaderName: HashMap<String?, String>? = null

    fun addStat(s: StatConfig) {
        statHash[s.statName] = s
    }

    fun addGraph(s: GraphConfig) {
        graphHash[s.name] = s
    }

    fun getStat(columns: Array<String>, firstdatacolumn: Int): String? {
        // this is called for each line of source file
        val s1 = Arrays.copyOfRange(columns, firstdatacolumn, columns.size)
        val header = java.lang.String.join(" ", *s1)

        // cache Mapping of HeaderStr to StatName - get StatHash more efficiently
        createCacheForMappingOfHeaderStr2StatName()
        val statName = mappingStatHeaderName!![header]
        return if (statName != null) {
            statHash[statName]!!.graphName
        } else {
            null
        }
    }

    private fun createCacheForMappingOfHeaderStr2StatName() {
        // do this only once - all Stats are known - create reverse maaping
        if (mappingStatHeaderName == null) {
            mappingStatHeaderName = HashMap()
            statHash.forEach { (k: String?, v: StatConfig) ->
                val statName = v.statName
                val graphHeader = v.headerStr
                mappingStatHeaderName!![graphHeader] = statName
            }
        }
    }

    fun getStat(statName: String): StatConfig? {
        if (statHash.isEmpty()) {
            return null
        }
        val result = arrayOf<StatConfig?>(null)
        statHash.keys.forEach(
            Consumer { item: String? ->
                val tmp = statHash[item]
                if (tmp!!.graphName == statName) {
                    result[0] = tmp
                }
            }
        )
        return result[0]
    }

    fun getGraphConfig(s: String): GraphConfig? {
        return if (graphHash.isEmpty()) {
            null
        } else graphHash[s]
    }
}
